db/migrate/20200129172534_add_epoch_version_release_arch_to_katello_installed_packages.rb

```diff
@@ -0,0 +1,44 @@
+class AddEpochVersionReleaseArchToKatelloInstalledPackages < ActiveRecord::Migration[5.2]
+  def up
+    add_column :katello_installed_packages, :nvrea, :string
+    add_column :katello_installed_packages, :epoch, :string
+    add_column :katello_installed_packages, :version, :string
+    add_column :katello_installed_packages, :release, :string
+    add_column :katello_installed_packages, :arch, :string
+
+    epoch_non_0 = ::Katello::Rpm.where.not(epoch: 0)
+
+    ::Katello::InstalledPackage.reset_column_information
+    ::Katello::InstalledPackage.all.each do |pkg|
+      epoch = epoch_non_0.where(nvra: pkg.nvra).first
+      if epoch.nil?
+        epoch = '0'
+      end
+
+      pkg.epoch = epoch
+
+      nvra = pkg.nvra
+      arch = nvra[nvra.rindex('.') + 1..-1]
+      nvra.sub('.' + arch, '')
+
+      tokens = nvra.split('-')
+      release = tokens[-1]
+      version = tokens[-2]
+
+      pkg.arch = arch
+      pkg.release = release
+      pkg.version = version
+
+      pkg.nvrea = "#{epoch}:#{pkg.name}-#{version}-#{release}.#{arch}"
+      pkg.save!
```

Comment:
i think saving these to an array and then using the bulk importing library would speed this up a lot

Reply:
Looking into this, for some reason, even if I do `::Katello::InstalledPackage.import(installed_packages, validate: false, on_duplicate_key_update: :all)` there are still errors, like `PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_katello_installed_packages_on_nvra"
DETAIL:  Key (nvra)=(cheetah-1.25.3-5.noarch) already exists.`

Reply:
Oh I see why, activerecord import's `on_duplicate_key_update` requires PostgreSQL 9.5+, but in dev we're on `"PostgreSQL 9.2.24 on x86_64-redhat-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-28), 64-bit"`